build_rules.gradle

```diff
@@ -1260,6 +1261,146 @@ ext.applyGrpcNature = {
 
 /*************************************************************************************************/
 
+ext.applyPortabilityNature = {
+  println "applyPortabilityNature with " + (it ? "$it" : "default configuration") + " for project $project.name"
+  applyJavaNature(failOnWarning: true, enableFindbugs: false, shadowClosure: {
+    // The relocation paths below specifically use the major version number of the dependency
+    // since we assume that packages following semantic versioning rules. For packages
+    // with a major version number that is 0 (implying that the dependency is unstable),
+    // we use the full version identifier in the relocated path (e.g. v0_minor_patch).
+    // For stable dependencies, if two or more packages use the same major version number,
```

Comment:
Thanks for writing this extensive documentation. This will be a useful reference for anyone that needs to maintain dependencies.

Could you define what you mean by "stable dependencies"; in contrast to those which change more frequently?

It may also be useful to talk about the implications of having multiple versions of a dependency. IMO shading is one of the more difficult-to-understand areas of the Beam codebase, so I welcome any documentation that helps explain how we use it. This may even deserve separate markdown doc or wiki page.

Just my 2cents. Feel free to disagree or defer to later.

Reply:
"stable" is likely anything that is 1.0.0 or greater unless the someone knows that versions aren't stable.
Unfortunately I don't believe there is a hard and fast rule and it all depends on the person doing the work to know what is stable. For example, most people would consider Jackson 2.x to be stable but in reality they do remove things and make backwards incompatible changes between minor versions (Jackson marks things deprecated and then remove after some number of minor versions) so we may find ourselves in cases where we need to be more granular then just the major version number. In other cases some libraries use completely different package namespaces between major versions which means that you don't even need to include the major version identifier.